docs: update operator imports in examples

jakovljevic-mladen opened the pull request:
**Description:**
As of RxJS version 7.2.0, most operators can be imported from `'rxjs'` export site. This PR updates code examples in most places. However, content from `*.md` files from `docs_app/content` folder is not updated because I'm still waiting for #6548 to be merged. Once that one gets merged, I will update files from `docs_app/content`.

Also, please forgive me for trying to be consistent and please try to ignore changes such as:
- removing brackets: `(x)` -> `x`
- removing trailing commas
- fixing whitespace issues
- adding missing `;`
- removing H3 headers (`###`) from examples - most of them are not present in most examples, so I removed almost all remaining ones
- replacing `"` with `'`
- and so on...

**Related issue (if exists):**
Closes #6650

benlesh commented on src/internal/ajax/ajax.ts:
```diff
@@ -12,16 +12,16 @@ export interface AjaxCreationMethod {
    * This is the most configurable option, and the basis for all other AJAX calls in the library.
    *
-   * ### Example
+   * ## Example
```
I honestly don't know what the implications are of changing this. @niklas-wortmann, do you have any idea how this might change the docs app output?

jakovljevic-mladen replied:
This changes from `<h3>` to `<h2>`. I did this in couple of files to align them with so many files from `/operators ` folder which all use `##`.

jakovljevic-mladen commented on src/internal/operators/max.ts:
```diff
@@ -9,34 +9,36 @@ import { isFunction } from '../util/isFunction';
- * import { of } from 'rxjs';
- * import { max } from 'rxjs/operators';
- *
- * interface Person {
```
TS can now infer type of objects passed to `of`, so I removed this interface entirely - I did this wherever I found similar interface declarations.

jakovljevic-mladen commented on src/internal/operators/publish.ts:
```diff
@@ -42,23 +42,23 @@ export function publish<T, O extends ObservableInput<any>>(selector: (shared: Ob
- * and subscribe.
+ *
+ * Make `source$` hot by applying `publish` operator, then merge each inner observable into a single one
+ * and subscribe
```
Most examples don't have dot (`.`) at the end of a sentence, so I removed it wherever there was one. I hate it, I'd rather have it...

jakovljevic-mladen commented on src/internal/operators/timeInterval.ts:
```diff
@@ -19,41 +18,33 @@ import { map } from './map';
- *     error: err => console.log(err),
- * });
- *
- * seconds.pipe(timeout(900))
```
Not sure how `timeout` came here, so I removed it.

jakovljevic-mladen commented on src/internal/operators/timestamp.ts:
```diff
@@ -8,24 +8,23 @@ import { map } from './map';
  * `{value: T, timestamp: R}`. The properties are generically typed. The `value` property contains the value
  * and type of the *source* observable. The `timestamp` is generated by the schedulers `now` function. By
- * default it uses the *async* scheduler which simply returns `Date.now()` (milliseconds since 1970/01/01
+ * default, it uses the `asyncScheduler` which simply returns `Date.now()` (milliseconds since 1970/01/01
```
I have a fix [for this](https://github.com/ReactiveX/rxjs/pull/6247#issuecomment-825491834) in my [other branch](https://github.com/jakovljevic-mladen/rxjs/tree/remove_unnecessary_images). If changes from that branch land to master, `asyncScheduler` will automatically become a link to the `asyncScheduler` docs.

benlesh commented:
I'm really sorry, @jakovljevic-mladen, this PR touches so many files in many different ways that It's going to take a very, very long time to review. There seem to be a number of unrelated changes all in one PR, and I'm having a hard time tracking them all mentally during review. (There are >1,400 lines changed and many for different reasons)

Do you think you can break this up into a PR for each type of change? (i.e. changing the docs "Examples"/"Example" headers, or  updating the operator imports)

benlesh commented:
I do want to say that I REALLY appreciate this work. It's important, and I'm tempted just to try to review this PR and push it through... but it's *because* it's so important that I want to make sure I'm able to review it properly and without confusion. If that makes sense.

jakovljevic-mladen commented:
No worries, I understand 🙂

I could do multiple PRs, no worries about that. But, if this works better for you: to transform a single commit from this PR to multiple commits, each having it's own purpose - I'd recommend that way though.

Please let me know @benlesh

benlesh commented:
@jakovljevic-mladen I'm indifferent, but please keep in mind that I'll need to cherry-pick these changes into the 8.x branch. If you think you can do on PR with N commits where each commit consolidates a set of similar changes, that's fine. I just don't want to be in a situation where I'm cherry-picking a ton of individual commits 😅 .  I mean, it seems like this could just be 4-5 commits, and that's fine if you want to do that.

jakovljevic-mladen commented:
@benlesh 

> If you think you can do on PR with N commits where each commit consolidates a set of similar changes, that's fine.

Yeah, I can. I already started working on it, should be finished soon. When it's done, doing the review commit-by-commit should be piece of cake 🙂 - as all changes from each commit will be related.

>it seems like this could just be 4-5 commits

There may be a bit more than 4-5 commits, but not much more, maybe 6-7 commits in total. And, I guess that they can all be squashed in the end for easier cherry-picking.

jakovljevic-mladen commented on src/internal/operators/timeInterval.ts:
````diff
@@ -23,8 +23,7 @@ import { map } from './map';
  * ```ts
- * import { interval } from "rxjs";
- * import { timeInterval, timeout } from "rxjs/operators";
+ * import { interval, timeInterval } from 'rxjs';
````
I dismissed `seconds.pipe(timeout(900))...` example from lines 36 through 40 in one of the later commits, this is why I don't have import for `timeout` here.

jakovljevic-mladen commented:
@benlesh, I managed to split this PR to multiple commits, almost all of them consisting of equally the same set of changes. I'm sorry for this large PR, I really hope this will make it a bit easier to review. Also, I'm sorry, but it it took 11 commits in the end, not 6-7 as I predicted.

Only one commit has some random changes that don't go to any other group, all others should have equally the same set of changes and all of them should be straightforward to review. Here's the list of them and some of my comments:

- [8b43097e](https://github.com/ReactiveX/rxjs/pull/6678/commits/8b43097ef744f1fd661a0b1bb447a4e4e91eb004) - this is the main commit that consist of the changes required by #6650. What I tried to do here is that I tried to list imported objects by the first appearance in the code example - so, if you see that I changed from e.g. `{ map, scan }` to `{ scan, map }`, this is most probably because `map` comes after `scan` in a pipe (e.g. `.pipe(scan(...), map(...))`. Also, all other imports are sorted this way - by the first appearance.
- [130d2960](https://github.com/ReactiveX/rxjs/pull/6678/commits/130d29603ec58893f7d875fe2d0ab95c479c875a) - this commit only updates `README.md` - I added changes before v7.2 and after v7.2.
- [27c19b5e](https://github.com/ReactiveX/rxjs/pull/6678/commits/27c19b5eb8eb6605ff1188820b2728c28eb56c9c) - replaced `typescript` with `ts` in code example headings - almost every other file in the whole repository uses `ts`.
- [99704ee9](https://github.com/ReactiveX/rxjs/pull/6678/commits/99704ee95438c0c30e88771d29aee830e35907fe) - replaced deprecated `async` with `asyncScheduler`.
- [82c27fa0](https://github.com/ReactiveX/rxjs/pull/6678/commits/82c27fa0eda128f49e5000590cf16229b6fbafbb) - fixed headings. Most headings in most operator files use `<h2>Example(s)<h2>` (markup synonym being `## Example(s)`), thus I replaced all H3 headings with H2. Also, most examples don't have dot (`.`) at the end of sentence, so I removed it from most of those that had it. I fixed whitespacing issues as well - mostly by adding some empty lines between a heading, a description text and a code snippet. Also, if there were many `## Example` headings in the same JSDoc, I removed all except the first one (the first one became `## Examples`). If some files had a description text prefixed with `###`, most files don't have this prefix, so I removed them as well.
- [0ae39ab8](https://github.com/ReactiveX/rxjs/pull/6678/commits/0ae39ab8bf88ab2659132735a7530ddd58a89c89) - fixed links that have this style: `{@link something}`. This is mostly being removal of the path, e.g. `index/interval` to just `interval`. If [the `{@link }` parser](https://github.com/ReactiveX/rxjs/blob/081ca2ba72/docs_app/tools/transforms/links-package/services/getLinkInfo.js#L68) can't find a valid link, it will throw an error. I tested all changes from this commit and they all worked. Also, I fixed some `@see` links, some of them were added, some (duplicates) were removed.
- [d95fed47](https://github.com/ReactiveX/rxjs/pull/6678/commits/d95fed477d265be802a8894b22bc5eae818606c1) - this one is the one that is a total mess. I tried to unify most things, I'm not sure that I managed to do it. Also, I tried to fix some typos and grammar issues. This is a list of some of changes:
  - used single quotes,
  - fixed whitespace issues,
  - deleted trailing commas,
  - removed extra parenthesis,
  - converted to arrow functions,
  - updated parameters to non-deprecated version of subscribe,
  - added missing `subscribe` call to the final observable from example,
  - removed `@prettier` JSDoc comments,
  - added missing `;`,
  - replaced `let` with `const`,
  - removed unused `interface Person` - TS can now infer data types passed to `of`/`from`,
  - fixed some TS issued with `event.target` by doing casting,
  - replaced errors passed to `throwError` and `throwIfEmpty` with arrow functions returning those errors.
- [7ca231a1](https://github.com/ReactiveX/rxjs/pull/6678/commits/7ca231a1e9d8a877947403006809f9a6c74e2545) - replaced `mapTo` and `concatMapTo` with `map` and `concatMap`, respectively, as there is a [high chance of those being deprecated](https://github.com/ReactiveX/rxjs/issues/6399) one day.
- [7ebbbfa0](https://github.com/ReactiveX/rxjs/pull/6678/commits/7ebbbfa015d1dc2a9d5706a25cc7edbf3eda959f) - `tap` now supports `subscribe` and `finalize` callbacks, so I simplified logging in some examples by utilizing those callbacks instead of using `defer` and/or `finalize`.
- [df209aff](https://github.com/ReactiveX/rxjs/pull/6678/commits/df209affa10acdd7c48d5366f7566cfdc2688b78) - I simplified `share` example by introducing `tap` and `take` - `tap` is used to remove `console.log` call in `map` and `take` to be able to document entire output.
- [fb9a0f7e](https://github.com/ReactiveX/rxjs/pull/6678/commits/fb9a0f7e0b49e612ad608d2cea7ebd657cb2e6ff) - fix TS error that occurs on this line: `const upperCase = materialized.pipe(dematerialize());`

I know that there is a huge number of changes unrelated to the main goal which is described in #6650. @benlesh, if you feel like having issues doing the review, please let me know, I can easily drop most commits - I'd leave only the first two as those are only related to #6650. I would cherry pick dropped commits to other branches and I could create separate PRs to address those issues.

samal-rasmussen commented:
What an absolutely monumental effort you've made in updating the imports. Bravo. I have one nit with it, which is that alphabetic sorting of import minimizes the amount of git merge conflicts. It's also an easier rule to follow in practice.

jakovljevic-mladen commented:
Thanks @samal-rasmussen! Also, thanks for your suggestion. I agree, alphabetic sorting is a better way, but it is usually handled by IDEs when it comes to the real code imports. Since these were JSDocs comments that I had to maintain manually, no IDE could help me with them, so it was much easier for me to keep track of all needed objects (and all unused imports - there were so many of them) than it would if I'd had to sort them alphabetically. Sorting alphabetically would require much more effort as I'd have to think what object comes before another one.

Though, I agree that some existing imports that didn't need to be changed could have been left unchanged, sorry about that.

benlesh commented on src/internal/operators/dematerialize.ts:
```diff
@@ -25,26 +25,26 @@ import { OperatorSubscriber } from './OperatorSubscriber';
- * const notifA = { kind: 'N', value: 'A' };
- * const notifB = { kind: 'N', value: 'B' };
- * const notifE = { kind: 'E', error: new TypeError('x.toUpperCase is not a function') }
+ * const notifA = <NextNotification<string>>{ kind: 'N', value: 'A' };
```
Please prefer `const notifA: NextNotification<string> = ` or `{} as NextNotification<string>`. We're trying to move away from bracket casts.

benlesh replied:
This is a "nit". We can fix it later.

jakovljevic-mladen replied:
Sure, I'll fix in another PR.

benlesh commented:
Thank you so much for all of this work, @jakovljevic-mladen! And thank you for breaking it into different commits. It was much easier to review.

jakovljevic-mladen commented:
Thank you @benlesh for having patience to review quite a massive PR.
